web/packages/shared/components/Notification/types.ts

```diff
@@ -16,20 +16,30 @@
  * along with this program.  If not, see <http://www.gnu.org/licenses/>.
  */
 
+import { Action } from 'design/Alert';
+import { IconProps } from 'design/Icon/Icon';
+import React from 'react';
+
+export type NotificationSeverity =
+  | 'info'
+  | 'warn'
+  | 'error'
+  | 'success'
+  | 'neutral';
+
 export interface NotificationItem {
   content: NotificationItemContent;
-  severity: 'info' | 'warn' | 'error';
+  severity: NotificationSeverity;
   id: string;
 }
 
 export type NotificationItemContent = string | NotificationItemObjectContent;
 
 export type NotificationItemObjectContent = {
-  title: string;
+  title?: string;
```

Comment:
I agree that we _might_ make it more strict, but think that this level of complexity is a bit of an overkill here. I'm a bit afraid that if we make it this complex, it's gonna bubble up the type hierarchy in unpredictable ways and cause errors with very confusing messages. For now, I'm fine with a risk of someone creating an empty notification. Am I missing something obvious here?

Reply:
It's true that the error messages generated by this type might end up being confusing. When I'm considering this tradeoff myself, I usually brush off the problem with error messages by believing that TypeScript will only improve its error messages over time [1], whereas we can enforce more safety through types right now.

When I encounter a confusing type error, I usually try to look up its source, which in this case would guide me towards a fix. The error message gets kind of better when we give names to individual types, but it's still far from perfect.

<details>
<summary>Types with names</summary>

```typescript
type TitleWithOptionalSubtitleAndDescription = {
  title: string;
  subtitle?: string;
  description?: string;
};

type DescriptionOnly = {
  description: string;
  title?: undefined;
  subtitle?: undefined;
};

export type NotificationItemObjectContent = (
  | TitleWithOptionalSubtitleAndDescription
  | DescriptionOnly
) & {
  list?: string[];
  icon?: React.ComponentType<IconProps>;
  action?: Action;
};
```

```
        <Notification
          …
            content: {
              subtitle: 'siemka',
            },
        />


web/packages/shared/components/Notification/Notification.story.tsx:69:13 - error TS2322: Type '{ subtitle: string; }' is not assignable to type 'NotificationItemContent'.
  Type '{ subtitle: string; }' is not assignable to type 'TitleWithOptionalSubtitleAndDescription & { list?: string[]; icon?: ComponentType<IconProps>; action?: Action; }'.
    Property 'title' is missing in type '{ subtitle: string; }' but required in type 'TitleWithOptionalSubtitleAndDescription'.

69             content: {
               ~~~~~~~

  web/packages/shared/components/Notification/types.ts:39:3
    39   title: string;
         ~~~~~
    'title' is declared here.
  web/packages/shared/components/Notification/types.ts:31:3
    31   content: NotificationItemContent;
         ~~~~~~~
    The expected type comes from property 'content' which is declared here on type 'NotificationItem'
```
</details>

It's fine if you want to go for optional types though. If we find that the lack of them poses a real problem, we can always add them in the future, which shouldn't require big changes.

[1]: Because in my mind it's their duty to catch up to where [Elm](https://elm-lang.org/news/compiler-errors-for-humans) and Rust were years ago. Which might not actually be true, hence "believing". 😶